internal/stacks/stackaddrs/removed.go

```diff
@@ -4,134 +4,389 @@
+	return ComponentInstance{
+		Component: Component{
+			Name: rf.Name,
+		},
+		Key: key,
+	}, diags
+}
+
 // ParseRemovedFrom parses the "from" attribute of a "removed" block in a
 // configuration and returns the address of the configuration object being
 // removed.
 //
 // In addition to the address, this function also returns a traversal that
 // represents the unparsed index within the from expression. Users can
 // optionally specify a specific index of a component to target.
-func ParseRemovedFrom(expr hcl.Expression) (Component, hcl.Expression, tfdiags.Diagnostics) {
-	var component Component
+func ParseRemovedFrom(expr hcl.Expression) (RemovedFrom, tfdiags.Diagnostics) {
 	var diags tfdiags.Diagnostics
 
-	traversal, index, hclDiags := exprToComponentTraversal(expr)
-	diags = diags.Append(hclDiags)
-	if hclDiags.HasErrors() {
-		return component, index, diags
-	}
+	removedFrom := RemovedFrom{}
 
-	if len(traversal) < 2 {
-		return component, index, diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid 'from' attribute",
-			Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
-			Subject:  expr.Range().Ptr(),
-		})
+	current, moreDiags := exprToComponentTraversal(expr)
+	diags = diags.Append(moreDiags)
+	if moreDiags.HasErrors() {
+		return RemovedFrom{}, diags
 	}
 
-	root, ok := traversal[0].(hcl.TraverseRoot)
-	if !ok || root.Name != "component" {
-		return component, index, diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid 'from' attribute",
-			Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
-			Subject:  expr.Range().Ptr(),
-		})
-	}
+	for current != nil {
 
-	name, ok := traversal[1].(hcl.TraverseAttr)
-	if !ok {
-		return component, index, diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid 'from' attribute",
-			Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
-			Subject:  expr.Range().Ptr(),
-		})
+		// we're going to parse the traversal in sets of 2-3 depending on
+		// the indices, so we'll check that now.
+		nextTraversal := current.Current
+
+		for len(nextTraversal) > 0 {
+			var currentTraversal hcl.Traversal
+			var indexExpr hcl.Expression
+
+			switch {
+			case len(nextTraversal) < 2:
+				// this is simply an error, we always need at least 2 values
+				// for either stack.name or component.name.
+				return RemovedFrom{}, diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid 'from' attribute",
+					Detail:   "The 'from' attribute must designate a component that has been removed, in the form `component.component_name` or `component.component_name[\"key\"].",
```

Comment:
nit, but I think this might provide a little more clarity

```suggestion
					Detail:   "The 'from' attribute must designate a component that has been removed, ending in the form `component.component_name` or `component.component_name[\"key\"].",
```

Reply:
Same with the other comment: https://github.com/hashicorp/terraform/pull/36814/files#diff-30de451b36b0d0667ee50d6839f85b679c73ec80856fc6613cd5de32edbafc25R131-R135.

I reword this is in a later PR, perhaps best to iterate on it there?